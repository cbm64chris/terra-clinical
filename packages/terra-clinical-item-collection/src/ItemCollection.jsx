import React from 'react';
import PropTypes from 'prop-types';
import ResponsiveElement from 'terra-responsive-element';
import 'terra-base/lib/baseStyles';

import Item from './Item';
import Header from './Header';
import ListView from './_ListView';
import TableView from './_TableView';
import './ItemCollection.scss';

const propTypes = {
  /**
   * The breakpoint to switch from a table view to a list view. Breakpoint options are tiny, small, medium, large, or huge.
   */
  breakpoint: PropTypes.oneOf(['tiny', 'small', 'medium', 'large', 'huge']),
  /**
   * The items that will flex between either a list item or a table row.
   */
  children: PropTypes.node.isRequired,
  /**
   * A callback event that will be triggered when selection state changes.
   */
  onChange: PropTypes.func,
  /**
   * The styles to spread to the list. List style options are isDivided and hasChevrons.
   */
  listStyles: PropTypes.shape({
    isDivided: PropTypes.bool,
    hasChevrons: PropTypes.bool,
  }),
  /**
   * The styles to spread to the table. Table style options are isPadded and isStriped.
   */
  tableStyles: PropTypes.shape({
    isPadded: PropTypes.bool,
    isStriped: PropTypes.bool,
  }),
  /**
   * The elements expected to be displayed. If a child is missing an element, ItemCollection will allocate space to maintain the provided layout.
   */
  requiredElements: PropTypes.shape({
    startAccessoryRequired: PropTypes.bool,
    displaysRequired: PropTypes.Number,
    commentRequired: PropTypes.bool,
    endAccessoryRequired: PropTypes.bool,
  }),
};

const defaultProps = {
  breakpoint: 'small',
  onChange: undefined,
  listStyles: undefined,
  tableStyles: undefined,
  requiredElements: { startAccessoryRequired: true, displaysRequired: 8, commentRequired: true, endAccessoryRequired: true },
};

class ItemCollection extends React.Component {
  constructor(props) {
    super(props);
    this.handleOnChange = this.handleOnChange.bind(this);
  }

  handleOnChange(event, selectedIndex) {
    this.setState({ selectedIndex });
    if (this.props.onChange) {
      this.props.onChange(event);
    }
  }

  render() {
    const { children, breakpoint, requiredElements, listStyles, tableStyles } = this.props;

    const listDisplay = (<ListView onChange={this.handleOnChange} listStyles={listStyles} requiredElements={requiredElements}>{children}</ListView>);
    const tableDisplay = (<TableView onChange={this.handleOnChange} tableStyles={tableStyles} requiredElements={requiredElements}>{children}</TableView>);

    const breakpointDisplay = {};
    breakpointDisplay[breakpoint] = tableDisplay;

    return (
      <ResponsiveElement
        responsiveTo="parent"
        defaultElement={listDisplay}
        {...breakpointDisplay}
      />
    );
  }
}

ItemCollection.propTypes = propTypes;
ItemCollection.defaultProps = defaultProps;

// Item Collection Building Blocks
ItemCollection.Item = Item;
ItemCollection.Header = Header;
ItemCollection.Comment = Item.Comment;
ItemCollection.Display = Item.Display;

export default ItemCollection;
