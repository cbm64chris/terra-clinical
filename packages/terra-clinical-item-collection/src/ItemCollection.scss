:local {
  .column-accessory,
  .list-accessory {
    width: 40px;
  }

  .content-accessory {
    vertical-align: middle !important;
    width: 40px;
  }

  .content-display,
  .content-comment {
    word-wrap: break-word;
  }
}
