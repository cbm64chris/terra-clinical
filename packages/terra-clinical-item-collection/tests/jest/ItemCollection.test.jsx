import React from 'react';
import ItemCollection from '../../src/ItemCollection';

it('should render a default ItemCollection', () => {
  const itemCollection = (
    <ItemCollection>
      <ItemCollection.Item startAccessory={<p>S</p>} comment={<ItemCollection.Comment text="test comment" />} endAccessory={<p>E</p>}>
        <ItemCollection.Display text="Display 1" />
        <ItemCollection.Display text="Display 2" />
        <ItemCollection.Display text="Display 3" />
      </ItemCollection.Item>
    </ItemCollection>
  );
  const component = shallow(itemCollection);
  expect(component).toMatchSnapshot();
});

it('should render an ItemCollection with list styles', () => {
  const itemCollection = (
    <ItemCollection listStyles={{ isDivided: true, hasChevrons: true }}>
      <ItemCollection.Item startAccessory={<p>S</p>} comment={<ItemCollection.Comment text="test comment" />} endAccessory={<p>E</p>}>
        <ItemCollection.Display text="Display 1" />
        <ItemCollection.Display text="Display 2" />
        <ItemCollection.Display text="Display 3" />
      </ItemCollection.Item>
    </ItemCollection>
  );
  const component = shallow(itemCollection);
  expect(component).toMatchSnapshot();
});

it('should render an ItemCollection with table styles', () => {
  const itemCollection = (
    <ItemCollection tableStyles={{ isPadded: true, isStriped: true }}>
      <ItemCollection.Item startAccessory={<p>S</p>} comment={<ItemCollection.Comment text="test comment" />} endAccessory={<p>E</p>}>
        <ItemCollection.Display text="Display 1" />
        <ItemCollection.Display text="Display 2" />
        <ItemCollection.Display text="Display 3" />
      </ItemCollection.Item>
    </ItemCollection>
  );
  const component = shallow(itemCollection);
  expect(component).toMatchSnapshot();
});

it('should render an ItemCollection with a breakpoint', () => {
  const itemCollection = (
    <ItemCollection breapoint="tiny" >
      <ItemCollection.Item startAccessory={<p>S</p>} comment={<ItemCollection.Comment text="test comment" />} endAccessory={<p>E</p>}>
        <ItemCollection.Display text="Display 1" />
        <ItemCollection.Display text="Display 2" />
        <ItemCollection.Display text="Display 3" />
      </ItemCollection.Item>
    </ItemCollection>
  );
  const component = shallow(itemCollection);
  expect(component).toMatchSnapshot();
});

it('should render an ItemCollection with requiredElements', () => {
  const requiredElements = { startAccessoryRequired: true, displaysRequired: 5, commentRequired: true, endAccessoryRequired: false };
  const itemCollection = (
    <ItemCollection requiredElements={requiredElements}>
      <ItemCollection.Item startAccessory={<p>S</p>} comment={<ItemCollection.Comment text="test comment" />} endAccessory={<p>E</p>}>
        <ItemCollection.Display text="Display 1" />
      </ItemCollection.Item>
    </ItemCollection>
  );
  const component = shallow(itemCollection);
  expect(component).toMatchSnapshot();
});
