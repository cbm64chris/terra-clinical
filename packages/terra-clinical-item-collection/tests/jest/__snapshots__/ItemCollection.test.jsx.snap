// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render a default ItemCollection 1`] = `
<ResponsiveElement
  defaultElement={
    <ListView
      requiredElements={
        Object {
          "commentRequired": true,
          "displaysRequired": 8,
          "endAccessoryRequired": true,
          "startAccessoryRequired": true,
        }
      }
    >
      <Item
        comment={
          <ItemComment
            isTruncated={false}
            text="test comment"
          />
        }
        endAccessory={
          <p>
            E
          </p>
        }
        listItemStyles={Object {}}
        startAccessory={
          <p>
            S
          </p>
        }
        view="list"
      >
        <ItemDisplay
          isTruncated={false}
          text="Display 1"
        />
        <ItemDisplay
          isTruncated={false}
          text="Display 2"
        />
        <ItemDisplay
          isTruncated={false}
          text="Display 3"
        />
      </Item>
    </ListView>
  }
  responsiveTo="parent"
  small={
    <TableView
      requiredElements={
        Object {
          "commentRequired": true,
          "displaysRequired": 8,
          "endAccessoryRequired": true,
          "startAccessoryRequired": true,
        }
      }
    >
      <Item
        comment={
          <ItemComment
            isTruncated={false}
            text="test comment"
          />
        }
        endAccessory={
          <p>
            E
          </p>
        }
        listItemStyles={Object {}}
        startAccessory={
          <p>
            S
          </p>
        }
        view="list"
      >
        <ItemDisplay
          isTruncated={false}
          text="Display 1"
        />
        <ItemDisplay
          isTruncated={false}
          text="Display 2"
        />
        <ItemDisplay
          isTruncated={false}
          text="Display 3"
        />
      </Item>
    </TableView>
  }
/>
`;

exports[`should render an ItemCollection with a breakpoint 1`] = `
<ResponsiveElement
  defaultElement={
    <ListView
      requiredElements={
        Object {
          "commentRequired": true,
          "displaysRequired": 8,
          "endAccessoryRequired": true,
          "startAccessoryRequired": true,
        }
      }
    >
      <Item
        comment={
          <ItemComment
            isTruncated={false}
            text="test comment"
          />
        }
        endAccessory={
          <p>
            E
          </p>
        }
        listItemStyles={Object {}}
        startAccessory={
          <p>
            S
          </p>
        }
        view="list"
      >
        <ItemDisplay
          isTruncated={false}
          text="Display 1"
        />
        <ItemDisplay
          isTruncated={false}
          text="Display 2"
        />
        <ItemDisplay
          isTruncated={false}
          text="Display 3"
        />
      </Item>
    </ListView>
  }
  responsiveTo="parent"
  small={
    <TableView
      requiredElements={
        Object {
          "commentRequired": true,
          "displaysRequired": 8,
          "endAccessoryRequired": true,
          "startAccessoryRequired": true,
        }
      }
    >
      <Item
        comment={
          <ItemComment
            isTruncated={false}
            text="test comment"
          />
        }
        endAccessory={
          <p>
            E
          </p>
        }
        listItemStyles={Object {}}
        startAccessory={
          <p>
            S
          </p>
        }
        view="list"
      >
        <ItemDisplay
          isTruncated={false}
          text="Display 1"
        />
        <ItemDisplay
          isTruncated={false}
          text="Display 2"
        />
        <ItemDisplay
          isTruncated={false}
          text="Display 3"
        />
      </Item>
    </TableView>
  }
/>
`;

exports[`should render an ItemCollection with list styles 1`] = `
<ResponsiveElement
  defaultElement={
    <ListView
      listStyles={
        Object {
          "hasChevrons": true,
          "isDivided": true,
        }
      }
      requiredElements={
        Object {
          "commentRequired": true,
          "displaysRequired": 8,
          "endAccessoryRequired": true,
          "startAccessoryRequired": true,
        }
      }
    >
      <Item
        comment={
          <ItemComment
            isTruncated={false}
            text="test comment"
          />
        }
        endAccessory={
          <p>
            E
          </p>
        }
        listItemStyles={Object {}}
        startAccessory={
          <p>
            S
          </p>
        }
        view="list"
      >
        <ItemDisplay
          isTruncated={false}
          text="Display 1"
        />
        <ItemDisplay
          isTruncated={false}
          text="Display 2"
        />
        <ItemDisplay
          isTruncated={false}
          text="Display 3"
        />
      </Item>
    </ListView>
  }
  responsiveTo="parent"
  small={
    <TableView
      requiredElements={
        Object {
          "commentRequired": true,
          "displaysRequired": 8,
          "endAccessoryRequired": true,
          "startAccessoryRequired": true,
        }
      }
    >
      <Item
        comment={
          <ItemComment
            isTruncated={false}
            text="test comment"
          />
        }
        endAccessory={
          <p>
            E
          </p>
        }
        listItemStyles={Object {}}
        startAccessory={
          <p>
            S
          </p>
        }
        view="list"
      >
        <ItemDisplay
          isTruncated={false}
          text="Display 1"
        />
        <ItemDisplay
          isTruncated={false}
          text="Display 2"
        />
        <ItemDisplay
          isTruncated={false}
          text="Display 3"
        />
      </Item>
    </TableView>
  }
/>
`;

exports[`should render an ItemCollection with requiredElements 1`] = `
<ResponsiveElement
  defaultElement={
    <ListView
      requiredElements={
        Object {
          "commentRequired": true,
          "displaysRequired": 5,
          "endAccessoryRequired": false,
          "startAccessoryRequired": true,
        }
      }
    >
      <Item
        comment={
          <ItemComment
            isTruncated={false}
            text="test comment"
          />
        }
        endAccessory={
          <p>
            E
          </p>
        }
        listItemStyles={Object {}}
        startAccessory={
          <p>
            S
          </p>
        }
        view="list"
      >
        <ItemDisplay
          isTruncated={false}
          text="Display 1"
        />
      </Item>
    </ListView>
  }
  responsiveTo="parent"
  small={
    <TableView
      requiredElements={
        Object {
          "commentRequired": true,
          "displaysRequired": 5,
          "endAccessoryRequired": false,
          "startAccessoryRequired": true,
        }
      }
    >
      <Item
        comment={
          <ItemComment
            isTruncated={false}
            text="test comment"
          />
        }
        endAccessory={
          <p>
            E
          </p>
        }
        listItemStyles={Object {}}
        startAccessory={
          <p>
            S
          </p>
        }
        view="list"
      >
        <ItemDisplay
          isTruncated={false}
          text="Display 1"
        />
      </Item>
    </TableView>
  }
/>
`;

exports[`should render an ItemCollection with table styles 1`] = `
<ResponsiveElement
  defaultElement={
    <ListView
      requiredElements={
        Object {
          "commentRequired": true,
          "displaysRequired": 8,
          "endAccessoryRequired": true,
          "startAccessoryRequired": true,
        }
      }
    >
      <Item
        comment={
          <ItemComment
            isTruncated={false}
            text="test comment"
          />
        }
        endAccessory={
          <p>
            E
          </p>
        }
        listItemStyles={Object {}}
        startAccessory={
          <p>
            S
          </p>
        }
        view="list"
      >
        <ItemDisplay
          isTruncated={false}
          text="Display 1"
        />
        <ItemDisplay
          isTruncated={false}
          text="Display 2"
        />
        <ItemDisplay
          isTruncated={false}
          text="Display 3"
        />
      </Item>
    </ListView>
  }
  responsiveTo="parent"
  small={
    <TableView
      requiredElements={
        Object {
          "commentRequired": true,
          "displaysRequired": 8,
          "endAccessoryRequired": true,
          "startAccessoryRequired": true,
        }
      }
      tableStyles={
        Object {
          "isPadded": true,
          "isStriped": true,
        }
      }
    >
      <Item
        comment={
          <ItemComment
            isTruncated={false}
            text="test comment"
          />
        }
        endAccessory={
          <p>
            E
          </p>
        }
        listItemStyles={Object {}}
        startAccessory={
          <p>
            S
          </p>
        }
        view="list"
      >
        <ItemDisplay
          isTruncated={false}
          text="Display 1"
        />
        <ItemDisplay
          isTruncated={false}
          text="Display 2"
        />
        <ItemDisplay
          isTruncated={false}
          text="Display 3"
        />
      </Item>
    </TableView>
  }
/>
`;
